components/shell/auth/oauth/useOAuth.ts

```diff
@@ -0,0 +1,288 @@
+
+function useCallbackRef<T extends (this: any, ...args: any[]) => any>(
+  fn?: T,
+): (this: ThisParameterType<T>, ...args: Parameters<T>) => void | ReturnType<T>
+
+function useCallbackRef<T extends (this: any, ...args: any[]) => any>(fn?: T) {
+  const fnRef = useRef(fn)
+  useEffect(() => {
+    fnRef.current = fn
+  }, [fn])
+  return useCallback(function (
+    this: ThisParameterType<T>,
+    ...args: Parameters<T>
+  ): void | ReturnType<T> {
+    const { current } = fnRef
+    if (current) return current.call(this, ...args)
+  },
+  [])
+}
+
+type ClientOptions =
+  | { client: BrowserOAuthClient }
+  | Pick<
+      BrowserOAuthClientLoadOptions,
+      | 'clientId'
+      | 'handleResolver'
+      | 'responseMode'
+      | 'plcDirectoryUrl'
+      | 'fetch'
+    >
+  | Pick<
+      BrowserOAuthClientOptions,
+      | 'clientMetadata'
+      | 'handleResolver'
+      | 'responseMode'
+      | 'plcDirectoryUrl'
+      | 'fetch'
+    >
+
+function useOAuthClient(options: ClientOptions): null | BrowserOAuthClient
+function useOAuthClient(
+  options: Partial<
+    { client: BrowserOAuthClient } & BrowserOAuthClientLoadOptions &
+      BrowserOAuthClientOptions
+  >,
+) {
+  const {
+    client: clientInput,
+    clientId,
+    clientMetadata,
+    handleResolver,
+    responseMode,
+    plcDirectoryUrl,
+  } = options
+
+  const [client, setClient] = useState<null | BrowserOAuthClient>(
+    clientInput || null,
+  )
+  const fetch = useCallbackRef(options.fetch || globalThis.fetch)
+
+  useEffect(() => {
+    if (clientInput) {
+      setClient(clientInput)
+    } else if (!handleResolver) {
+      throw new TypeError('handleResolver is required')
+    } else {
+      const ac = new AbortController()
+      const { signal } = ac
+
+      setClient(null)
+
+      import('@atproto/oauth-client-browser').then(({ BrowserOAuthClient }) => {
+        if (signal.aborted) return
+        if (clientMetadata || !clientId) {
+          const client = new BrowserOAuthClient({
+            clientMetadata,
+            handleResolver,
+            responseMode,
+            plcDirectoryUrl,
+            fetch,
+          })
+          signal.addEventListener('abort', () => client.dispose(), {
+            once: true,
+          })
+          setClient(client)
+        } else {
+          BrowserOAuthClient.load({
+            clientId,
+            handleResolver,
+            responseMode,
+            plcDirectoryUrl,
+            fetch,
+            signal,
+          }).then(
+            (client) => {
+              if (!signal.aborted) {
+                signal.addEventListener('abort', () => client.dispose(), {
+                  once: true,
+                })
+                setClient(client)
+              } else {
+                client.dispose()
+              }
+            },
+            (err) => {
+              if (!signal.aborted) throw err
+            },
+          )
+        }
+      })
```

Comment:
This gets a bit complicated to handle the branching paths of loading client metadata vs. baking it in.  I assume we just use one of these paths in ozone— can we remove the other?

Reply:
Actually did that without even seeing your comment here 😅